bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java

```diff
@@ -57,10 +57,20 @@ private synchronized void refreshZoneList() {
             Long homeId = homeHandler.getHomeId();
             if (homeId != null) {
                 logger.debug("Fetching (battery state) zone list for HomeId {}", homeId);
+                zoneList.clear();
                 try {
-                    zoneList = homeHandler.getApi().listZones(homeId);
+                    for (Zone zone : homeHandler.getApi().listZones(homeId)) {
```

Comment:
What's the idea here? If any `batteryState` is low, set the `zoneList` to `OFF` otherwise to `ON`?

```
homeHandler.getApi().listZones(homeId).forEach(zone -> {
    try {
        boolean batteryLow = !zone.getDevices().stream().map(ControlDevice::getBatteryState).anyMatch(s->!("NORMAL".equals(s)));
        zoneList.put(Long:valueOf(zone.getId()), OnOffType.from(batteryLow));
    } catch (IOException | ApiException e) {
        logger.debug("Fetch (battery state) zone list exception", e); 
    }
}
```

This also allows to process other zones if one zone failed (instead of before where processing stops when the first error occurs.

Reply:
I think I need to give you some background here..

In a tado system the Bridge Thing is the home "HomeId", and within each Home, there are Things for rooms "ZoneId" (for example Living Room Heating). In the majority of cases, each Zone has a single radiator control valve device in it. But the tado system does also allow for each Zone to have more than one radiator control valve; and in that case the two (or more) radiator control valves are slaved together under the single Zone Thing.
```
HomeId = "my address"
├── ZoneId = "Kitchen"
│   ├── DeviceID = "radiator control valve A"
├── ZoneId = "Living Room"
     ├── DeviceID = "radiator control valve B"
     ├── DeviceID = "radiator control valve C"
```
In the binding, the main API interface is a JSON request to http://something/something/ZoneId=J which fetches Channel information for the Zone like room temperature, set temperature, heating state. The binding polls this JSON request according to the refresh Interval defined in the Configuration Parameters.

HOWEVER, for the low battery alarm -- unlike the other Channels -- the checking for this alarm is NOT done at Zone level but rather at each radiator control valve device level. So for each device there is a separate JSON request to the API as http://something/something/DeviceId=K which reads the low battery alarm state.

The radiator control valve devices are not modelled as Things in the binding. So to bring their low battery alarm state into openHAB it is necessary to poll each device, and amalgamate these alarm values in the common Zone low battery battery alarm Channel (by OR'ing across all the radiator control valve devices in that Zone).

Furthermore, in the previous build of this binding, the JSON calls for the low battery alarm were not being made according to the polling refresh Interval, but were instead being made immediately every time that the binding requested an update. So this caused far too many unnecessary API calls -- particularly since this Channel only goes into alarm roughly once in 6 months.

So in this new build of the binding, I have done two things: a) the polling of JSON request to the radiator control valve devices for their low battery alarm state is no longer made on every refresh request, and b) there is now a Map table that maps between the Zone Ids and the amalgamated low battery alarm states of the radiator control valve devices within that Zone.

Reply:
I think that is exactly what my code-snippet is doing. It's only another way of doing the same thing.

Reply:
Resolved. See comments below..